src/main/java/org/treadmill/zk/plugin/utils/LdapQuery.java

```diff
@@ -1,30 +1,60 @@
 package org.treadmill.zk.plugin.utils;
 
 import com.google.inject.Inject;
+import com.google.inject.Provider;
+import com.google.inject.Singleton;
 import com.unboundid.ldap.sdk.*;
 import org.slf4j.Logger;
 
 import java.io.IOException;
-import java.util.Collection;
 import java.util.List;
+import java.util.Set;
 
+import static java.util.Collections.emptySet;
+import static java.util.stream.Collectors.toSet;
 import static org.slf4j.LoggerFactory.getLogger;
+import static org.treadmill.zk.plugin.utils.Configuration.get;
 
+@Singleton
 public class LdapQuery {
 
-  private final LDAPInterface connection;
-
   private static final Logger logger = getLogger(LdapQuery.class);
+  Provider<LDAPInterface> ldapProvider;
 
   @Inject
-  public LdapQuery(LDAPConnection connection) {
-    this.connection = connection;
+  public LdapQuery(Provider<LDAPInterface> ldapProvider) throws IOException {
+    this.ldapProvider = ldapProvider;
   }
 
-  public Collection<Attribute> getAttributes(String baseDN, String filter) throws LDAPException, IOException {
+  public Set<String> searchServers() throws IOException, LDAPSearchException {
+    String baseDN = get("base_dn_for_server_role");
+    String filter = get("filter_for_server_role");
+    List<SearchResultEntry> searchEntries = ldapProvider.get()
+      .search(baseDN, SearchScope.ONE, filter).getSearchEntries();
+
+    if (searchEntries == null) return emptySet();
+
+    logger.info("found {} attributes for baseDN={}, filter={}", searchEntries.size(), baseDN, filter);
 
-    List<SearchResultEntry> searchEntries = connection.search(baseDN, SearchScope.ONE, filter).getSearchEntries();
-    logger.info("Found {} attributes for baseDN={} and filter={}", searchEntries.size(), baseDN, filter);
-    return searchEntries.get(0).getAttributes();
+    return searchEntries.stream()
+      .map(s -> s.getAttribute("server").getValue())
+      .collect(toSet());
   }
+
+  public Set<String> searchAdmins() throws IOException, LDAPSearchException {
+    String baseDN = get("base_dn_for_admin_role");
+    String filter = get("filter_for_admin_role");
+    SearchResultEntry searchResultEntry = ldapProvider.get()
+      .searchForEntry(baseDN, SearchScope.ONE, filter);
+
+    if (searchResultEntry == null) return emptySet();
+
+    logger.info("found {} attributes for baseDN={}, filter={}",
+      searchResultEntry.getAttributes().size(), baseDN, filter);
+
+    return searchResultEntry.getAttributes().stream()
```

Comment:
I am also thinking of the case were an admin user would/could have principals in multiple realms.  Say the FreeIPA Treadmill realm and some AD for their desktop from where they issue commands.

What would be the setup here? Would we configure TM LDAP with multiple "admin", each for one of these principals?